controller/snapshot_controller.go

```diff
@@ -0,0 +1,585 @@
+package controller
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	"github.com/longhorn/longhorn-manager/engineapi"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+)
+
+type SnapshotController struct {
+	*baseController
+
+	// which namespace controller is running with
+	namespace string
+	// use as the OwnerID of the controller
+	controllerID string
+
+	kubeClient    clientset.Interface
+	eventRecorder record.EventRecorder
+
+	ds                     *datastore.DataStore
+	cacheSyncs             []cache.InformerSynced
+	engineClientCollection engineapi.EngineClientCollection
+
+	monitor *SnapshotMonitor
+}
+
+type SnapshotCache struct {
+	logger logrus.FieldLogger
+
+	namespace string
+	ds        *datastore.DataStore
+
+	controllerID string
+
+	mutex         sync.RWMutex
+	snapshotStore map[string]*longhorn.Snapshot
+}
+
+type SnapshotMonitor struct {
+	logger logrus.FieldLogger
+
+	namespace string
+	ds        *datastore.DataStore
+
+	controllerID string
+
+	mutex         sync.RWMutex
+	snapshotStore map[string]*longhorn.Snapshot
+}
+
+func NewSnapshotController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	namespace string,
+	controllerID string,
+	engineClientCollection engineapi.EngineClientCollection) *SnapshotController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	sc := &SnapshotController{
+		baseController: newBaseController("longhorn-snapshot", logger),
+
+		namespace:              namespace,
+		controllerID:           controllerID,
+		kubeClient:             kubeClient,
+		eventRecorder:          eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: "longhorn-snapshot-controller"}),
+		ds:                     ds,
+		engineClientCollection: engineClientCollection,
+	}
+
+	sc.monitor = &SnapshotMonitor{
```

Comment:
Isn't the snapshot monitor used?

Reply:
Great catch. No, they are not being used. Removed. Thank you